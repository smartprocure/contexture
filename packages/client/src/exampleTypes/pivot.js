import _ from 'lodash/fp'
import { toJS } from 'mobx'

let getKey = x => x.keyAsString || x.key

// Similar to Tree.lookup but path is a drilldown which uses keyAsString or key
let resultsForDrilldown = (type, drilldown, results) => {
  if (_.isEmpty(drilldown) || !results) return results

  let key = _.first(drilldown)
  let groups = _.get(type, results)
  let match = _.find(node => getKey(node) === key, groups)

  return resultsForDrilldown(type, drilldown.slice(1), match)
}

let previouslyLoadedKeys = (expansion, expansions) =>
  _.flow(
    _.filter(
      ({ type, drilldown, loaded }) =>
        type === expansion.type &&
        _.isEqual(drilldown, expansion.drilldown) &&
        loaded
    ),
    _.flatMap('loaded')
  )(expansions)

let getResultKeys = (expansion, node, results) => {
  let groupType = expansion.type
  let expansions = node.expansions
  let drilldown = _.getOr([], 'drilldown', expansion)
  let drilldownResults = resultsForDrilldown(groupType, drilldown, results)
  let loadedKeys = _.map(getKey, _.get(groupType, drilldownResults))
  return _.flow(
    _.without(previouslyLoadedKeys(expansion, expansions)),
    _.take(_.getOr(Infinity, [groupType, drilldown.length, 'size'], node))
  )(loadedKeys)
}

let mergeResults = _.mergeWith((current, additional, prop) => {
  if (prop === 'columns' || prop === 'rows') {
    return _.flow(
      _.map(_.keyBy('key')),
      _.spread(mergeResults),
      _.values
    )([current, additional])
  } else if (_.isArray(additional)) return additional
})

// Resetting the expansions when the pivot node is changed
// allows to return expected root results instead of merging result
// EX: changing the columns or rows config was not returning the new results
let resetExpansions = (extend, node) => {
  extend(node, {
    expansions: [],
  })
}

// Resetting the row expansions when the sorting is changed
let resetExpandedRows = (extend, node) => {
  extend(node, {
    expansions: _.filter({ type: 'columns' }, node.expansions),
  })
}

let expand = (tree, path, type, drilldown) => {
  path = toJS(path)
  drilldown = toJS(drilldown)
  let node = toJS(tree.getNode(path))
  let expansions = node.expansions
  let results = node.context.results

  // adding values for initial root level expansions
  if (!_.find({ type: 'columns' }, expansions)) {
    let columnsExpansion = {
      type: 'columns',
      drilldown: [],
    }
    columnsExpansion.loaded = getResultKeys(columnsExpansion, node, results)
    expansions.push(columnsExpansion)
  }
  // adding values for initial root level expansions
  if (!_.find({ type: 'rows' }, expansions)) {
    let rowsExpansion = {
      type: 'rows',
      drilldown: [],
    }
    rowsExpansion.loaded = getResultKeys(rowsExpansion, node, results)
    expansions.push(rowsExpansion)
  }

  tree.mutate(path, {
    expansions: [
      ...expansions,
      {
        type,
        drilldown,
        loaded: false,
      },
    ],
  })
}
let collapse = (tree, path, type, drilldown) => {
  path = toJS(path)
  drilldown = toJS(drilldown)
  let node = tree.getNode(path)
  let results = toJS(_.get('context.results', node))
  let drilldownResults = resultsForDrilldown(type, drilldown, results)

  // removing expansions under this drilldown level
  node.expansions = _.filter(
    expansion =>
      expansion.type !== type ||
      !_.isEqual(
        // expantion.drilldown is not a child of drilldown
        _.take(drilldown.length, expansion.drilldown),
        _.toArray(drilldown)
      )
  )(node.expansions)

  // removing collapsed rows or columns from results
  drilldownResults[type] = undefined
  // triggering observer update
  tree.extend(node, { context: { results } })
}

export default {
  validate: context =>
    _.every(
      ({ type, ranges, percents }) =>
        (type !== 'numberRanges' && type !== 'percentiles') ||
        (type === 'numberRanges' && ranges.length > 0) ||
        (type === 'percentiles' && percents.length > 0),
      _.concat(context.columns, context.rows)
    ),
  reactors: {
    columns: 'self',
    rows: 'self',
    values: 'self',
    expansions: 'self',
    filters: 'others',
    sort: 'self',
  },
  defaults: {
    columns: [],
    rows: [],
    values: [],
    filters: [],
    sort: {},
    showCounts: false,
    expanded: {
      columns: false,
      rows: false,
    },
    expand,
    collapse,
    expansions: [
      /*
     {
        type: 'columns',
        drilldown: [],
        loaded: [ 'x', 'y', 'z' ],
      },
     {
        type: 'rows',
        drilldown: [],
        loaded: [ 'a', 'b', 'c' ],
      },
      {
        type: 'rows',
        drilldown: [],
        loaded: false,
        // falsy loaded to request more results
        // will skip [ 'a', 'b', 'c' ] automatically
      },
     */
    ],
    context: {
      results: {},
    },
  },
  onDispatch(event, extend) {
    let { type, node, previous, value } = event

    if (type !== 'mutate') return

    // if sorting is changed we are preserving expanded columns
    if (_.has('sort', value)) return resetExpandedRows(extend, node)

    if (_.has('expansions', value)) return

    // if anything else about node configuration is changed resetting pages

    resetExpansions(extend, node, previous)
  },
  // Resetting the expansions when the tree is changed
  // allows to return expected root results instead of nested drilldown
  // EX: criteria filters didn't work properly when drilldown was applied
  onUpdateByOthers(node, extend) {
    resetExpansions(extend, node, node)
  },
  shouldMergeResponse: node => !_.isEmpty(node.expansions),
  mergeResponse(node, response, extend, snapshot) {
    let nextEmptyExpansion = () =>
      _.find(({ loaded }) => !loaded, node.expansions)

    let expansion

    while ((expansion = nextEmptyExpansion())) {
      // adding values to loaded expansion
      expansion.loaded = getResultKeys(
        expansion,
        node,
        response.context.results
      )
    }

    let context = mergeResults(snapshot(node.context), response.context)

    // Write on the node
    extend(node, { context })
  },
}
