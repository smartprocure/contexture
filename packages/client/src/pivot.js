import _ from 'lodash/fp'
import { deepMultiTransformOn } from './util/futil'
import F from 'futil'

let getKey = x => x.keyAsString || x.key

// Similar to Tree.lookup but path is a drilldown which uses keyAsString or key
let resultsForDrilldown = (type, path, results) => {
  if (_.isEmpty(path) || !results) return results

  let key = _.first(path)
  let groups = _.get(type, results)
  let match = _.find(node => getKey(node) === key, groups)
  return resultsForDrilldown(type, path.slice(1), match)
}

let someNotEmpty = _.some(_.negate(_.isEmpty))

// Getting previous drilldown keys and adding include
// based on results and skip
let getDrilldownWithInclude = (node, previous, drillType) => {
  let page = _.get(['pagination', drillType], previous)

  if (page) {
    // adding root level include for the first page only if expanded is empty yet
    // allows us to remember keys of existing root level rows/columns
    if (
      _.isEmpty(page.drilldown) &&
      _.isEmpty(page.skip) &&
      !_.isEmpty(_.get(['pagination', drillType, 'expanded'], previous))
    )
      return false

    // getting existing results for the drilldown level
    let results = resultsForDrilldown(
      drillType,
      page.drilldown,
      _.get('context.results', node)
    )

    return {
      drilldown: page.drilldown,
      // getting existing row/column keys without skipped ones
      include: _.flow(
        _.get(drillType),
        _.map(getKey),
        _.without(page.skip)
      )(results),
    }
  }

  return false
}

let addDrilldownToExpanded = (extend, node, previous, value) => {
  let prevRowDrill = _.get('pagination.rows.drilldown', previous)
  let prevColumnDrill = _.get('pagination.columns.drilldown', previous)
  let isRowPagination = _.has('pagination.rows', value)

  // Previous expanded with previous drilldown added
  let getExpanded = drillType =>
    _.compact([
      ..._.getOr([], ['pagination', drillType, 'expanded'], previous),
      getDrilldownWithInclude(node, previous, drillType),
    ])

  // Preserving previous pagination entries in the expanded prop
  extend(node, {
    pagination: {
      columns: {
        drilldown: prevColumnDrill && [],
        skip: [],
        ...(!isRowPagination && _.get('pagination.columns', value)),
        expanded: getExpanded('columns'),
      },
      rows: {
        drilldown: prevRowDrill && [],
        skip: [],
        ...(isRowPagination && _.get('pagination.rows', value)),
        expanded: getExpanded('rows'),
      },
    },
  })
}

// Resetting the pagination when the pivot node is changed
// allows to return expected root results instead of merging result
// EX: changing the columns or rows config was not returning the new results
let resetPagination = (extend, node, previous) => {
  // If drilldown mode is enabled for columns or rows
  let prevRowDrill = _.get('pagination.rows.drilldown', previous)
  let prevColumnDrill = _.get('pagination.columns.drilldown', previous)

  extend(node, {
    pagination: {
      columns: {
        drilldown: prevColumnDrill && [],
        skip: [],
        expanded: [],
      },
      rows: {
        drilldown: prevRowDrill && [],
        skip: [],
        expanded: [],
      },
    },
  })
}

// Resetting the row pagination when the sorting is changed
let resetExpandedRows = (extend, node) => {
  let rowDrill = _.get('pagination.rows.drilldown', node)
  let columnDrill = _.get('pagination.columns.drilldown', node)
  let expandedColumns = _.get('pagination.columns.expanded', node)

  extend(node, {
    pagination: {
      columns: {
        drilldown: columnDrill && [],
        skip: [],
        expanded: expandedColumns,
      },
      rows: {
        drilldown: rowDrill && [],
        skip: [],
        expanded: [],
      },
    },
  })
}

export default {
  validate: context =>
    _.every(
      ({ type, ranges, percents }) =>
        (type !== 'numberRanges' && type !== 'percentiles') ||
        (type === 'numberRanges' && ranges.length > 0) ||
        (type === 'percentiles' && percents.length > 0),
      _.concat(context.columns, context.rows)
    ),
  reactors: {
    columns: 'self',
    rows: 'self',
    values: 'self',
    pagination: 'self',
    filters: 'others',
    sort: 'self',
  },
  defaults: {
    columns: [],
    rows: [],
    values: [],
    filters: [],
    sort: {},
    pagination: {
      columns: {
        drilldown: null,
        skip: [],
        expanded: [],
      },
      rows: {
        drilldown: null,
        skip: [],
        expanded: [],
      },
    },
    showCounts: false,
    context: {
      results: [],
    },
  },
  onDispatch(event, extend) {
    let { type, node, previous, value } = event

    if (type !== 'mutate') return

    if (_.has('pagination.columns', value) || _.has('pagination.rows', value))
      // If mutation is a pagination
      addDrilldownToExpanded(extend, node, previous, value)
    else if (_.has('sort', value)) {
      // if sorting is changes we are preserving expanded columns
      addDrilldownToExpanded(extend, node, previous, value)
      resetExpandedRows(extend, node)
    } else
      // If node configuration is changes disable mergeResponse
      resetPagination(extend, node, previous)
  },
  // Resetting the pagination when the tree is changed
  // allows to return expected root results instead of nested drilldown
  // EX: criteria filters didn't work properly when drilldown was applied
  onUpdateByOthers(node, extend) {
    resetPagination(extend, node, node)
  },
  shouldMergeResponse: node =>
    // checking for presence of drilldown, skip, expanded in pagination
    someNotEmpty(
      _.map(_.get(_, node), [
        'pagination.columns.drilldown',
        'pagination.columns.skip',
        'pagination.rows.drilldown',
        'pagination.rows.skip',
      ])
    ),
  mergeResponse(node, response, extend, snapshot) {
    // Convert response rows and columns to objects for easy merges
    let groupsToObjects = deepMultiTransformOn(
      ['rows', 'columns'],
      groupsToObjects => _.flow(_.map(groupsToObjects), _.keyBy('key'))
    )
    // Convert rows and columns back to arrays
    let groupsToArrays = deepMultiTransformOn(
      ['rows', 'columns'],
      groupsToArrays => _.flow(_.values, _.map(groupsToArrays))
    )

    // `snapshot` here is to solve a mobx issue
    let nodeGroups = groupsToObjects(snapshot(node.context.results))
    let responseGroups = groupsToObjects(response.context.results)

    // Easy merge now that we can merge by row key
    let results = F.mergeAllArrays([nodeGroups, responseGroups])

    let context = { results: groupsToArrays(results) }

    // Write on the node
    extend(node, { context })
  },
}
