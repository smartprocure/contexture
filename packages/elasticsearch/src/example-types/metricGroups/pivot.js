let F = require('futil')
let _ = require('lodash/fp')
let { getStats } = require('./stats')
let { getField } = require('../../utils/fields')
let types = require('../../../src/example-types')
let { transmuteTree } = require('../../utils/futil')
let { simplifyBucket } = require('../../utils/elasticDSL')

let lookupTypeProp = (def, prop, type) =>
  _.getOr(def, `${type}GroupStats.${prop}`, types)

// PivotTable -> Query:
//  rows -> columns -> values
//  (same) values for ALL group levels
// Rows/Columns are buckets, values are metrics

let aggsForValues = (node, schema) =>
  _.flow(
    // Add `pivotMetric-` to auto keys so we can skip camelCasing it in the response
    _.keyBy(
      ({ key, type, field }) =>
        key || F.compactJoin('-', ['pivotMetric', type, field])
    ),
    // This is a very interesting lint error - while key is not used in the function body, it is important.
    // Omitting key here would include it in the props spread and then pass it along as part of the body of the ES aggregation.
    // eslint-disable-next-line
    _.mapValues(({ key, type, field, ...props }) => ({
      [_.snakeCase(type)]: {
        ...props,
        ...(field && { field: getField(schema, field) }),
      },
    }))
  )(node)

let buildQuery = async (node, schema, getStats) => {
  let drilldowns = node.drilldown || []
  // Don't consider deeper levels than +1 the current drilldown
  // This allows avoiding expansion until ready
  // Opt out with falsey drilldown
  let groups = node.drilldown
    ? _.take(_.size(node.drilldown) + 1, node.groups)
    : node.groups

  let statsAggBlob = { aggs: aggsForValues(node.values, schema) }
  let query = await _.reduce(
    async (children, group) => {
      // Subtotals calculates metrics at each group level, not needed if flattening or in chart
      // Support for per group stats could also be added here - merge on another stats agg blob to children based on group.stats/statsField or group.values
      if (node.subtotals) children = _.merge(await children, statsAggBlob)
      let { type } = group
      let buildGroupQuery = lookupTypeProp(_.identity, 'buildGroupQuery', type)
      return buildGroupQuery(group, await children, schema, getStats)
    },
    statsAggBlob,
    // TODO: Also consider rows + columns -> groups:
    // _.reverse([...(node.rows || []), ...(node.columns || []), ...(node.groups || [])])
    // _.reverse(_.concat(node.rows, node.columns, node.groups)))
    _.reverse(groups)
  )

  let filters = _.compact(
    await Promise.all(
      F.mapIndexed((group, i) => {
        let filter = lookupTypeProp(_.stubFalse, 'drilldown', group.type)
        // stamp on drilldown from root if applicable
        let drilldown = drilldowns[i] || group.drilldown
        return drilldown && filter({ drilldown, ...group }, schema, getStats)
      }, groups)
    )
  )
  if (!_.isEmpty(filters))
    query = {
      aggs: {
        pivotFilter: {
          filter: { bool: { must: filters } },
          ...query,
        },
      },
    }

  return query
}

// TODO: instead of groupN, maybe look at query to say something more valuable?
//  e.g. => node.groups[n].field + ' ' + node.groups[n].type (aka 'Organization.NameState fieldValues'), maybe customized per type?
//      eg. `PO.IssuedDate month dateInterval`
// maybe client side since it's schema label driven?
let bucketToGroupN = (bucket, n) => ({
  [`group${n}`]: bucket.keyAsString || bucket.key,
})
let Tree = F.tree(_.get('groups'))
let flattenGroups = Tree.leavesBy((node, index, parents) => ({
  ...node,
  // Add groupN keys
  ..._.mergeAll(
    F.mapIndexed(
      bucketToGroupN,
      // dropping root parent (since that is just the aggs top level - would change if we add "totals" to the subtotals)
      // might also change based on what we pass in (e.g. pass in aggregations.groups?)
      _.reverse([node, ..._.dropRight(1, parents)])
    )
  ),
}))

let defaultGetGroups = _.get('groups.buckets')
// This captures everything but encodes types specific knowledge:
// let defaultGetGroups = aggs =>
//   F.when(
//     _.isPlainObject,
//     F.unkeyBy('key'),
//     (aggs.valueFilter || aggs).groups.buckets
//   )
let ensureGroups = node => {
  if (!_.isArray(node.groups)) node.groups = []
}
let processResponse = (response, node = {}) => {
  // Don't consider deeper levels than +1 the current drilldown
  // This allows avoiding expansion until ready
  // Opt out with falsey drilldown
  let groups = node.drilldown
    ? _.take(_.size(node.drilldown) + 1, node.groups || [])
    : (node.groups || [])

  // Traversing the ES response utilizes type specific methods looked up by matching the depth with node.groups
  let traverseSource = (x, i, parents = []) => {
    let depth = parents.length
    let { type } = groups[depth] || {}
    let traverse = lookupTypeProp(defaultGetGroups, 'getGroups', type)
    return traverse(x)
  }

  // Goal here is to map the tree from one structure to another
  // goal is to keep _nodes_ the same, but write back with different (dynamic) traversal
  //   e.g. valuefilter.groups.buckets -> groups, groups.buckets -> groups
  let simplifyTree = transmuteTree(traverseSource, Tree.traverse, ensureGroups)
  let results = simplifyTree(
    simplifyBucket,
    F.getOrReturn('pivotFilter', response.aggregations)
  )
  return { results: node.flatten ? flattenGroups(results) : results.groups }
}

let pivot = {
  aggsForValues,
  buildQuery,
  processResponse,
  validContext: node => node.groups.length && node.values.length,
  // TODO: unify this with groupStatsUtil - the general pipeline is the same conceptually
  async result(node, search, schema) {
    let query = await buildQuery(node, schema, getStats(search))
    // console.log(JSON.stringify({ query }, 0, 2))
    let response = await search(query)
    // console.log(JSON.stringify({ response }, 0, 2))
    let result = processResponse(response, node)
    // console.log(JSON.stringify({ result }, 0, 2))
    return result
  },
}
module.exports = pivot
