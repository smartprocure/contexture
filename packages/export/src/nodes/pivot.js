import _ from 'lodash/fp.js'
import F from 'futil'
import ContextureClient from 'contexture-client'
import {
  getKey,
  resultsForDrilldown,
} from 'contexture-client/exampleTypes/pivot.js'
import { andGroup, runWith, setFilterOnly } from '../utils.js'

// TODO move to shared folder
let getPageSize = (grouping) => {
  let pageSizeGetters = {
    fieldValues: _.getOr(10, 'size'),
  }
  return _.getOr(_.noop, grouping.type, pageSizeGetters)(grouping)
}

export let getGroupingSize = (node, groupingType, cardinalityResult, exportAllPages) => {
  let getNested = _.get(groupingType)
  let groupingCardinality = 1 // starting with 1 for the total column/row

  F.walk(getNested)((groupResult, index, parents) => {
    let groupCardinality = _.getOr(
      _.size(getNested(groupResult)),
      `${groupingType}GroupCount`,
      groupResult
    )
    if (!groupCardinality) return

    let grouping = _.get([groupingType, _.size(parents)], node)

    if (!exportAllPages)
      groupCardinality = _.min([groupCardinality, getPageSize(grouping)])

    // TODO else approximate full cardinality by using the average size of the first page of each grouping

    groupingCardinality = groupingCardinality + groupCardinality
  })(cardinalityResult)

  return groupingCardinality
}

export default async ({ service, tree, exportAllPages, ...node }) => {
  let run = (node) => runWith(service, tree, node)

  let getContextureClient = (tree, node) =>
    ContextureClient(
      {
        service,
        debounce: 0,
      },
      andGroup(setFilterOnly(tree), node)
    )

  let getGroupingResult = async (type) => {
    let groups = _.cloneDeep(_.get(type, node))
    let lastGroup = _.last(groups)
    if (_.isEmpty(groups)) return {}

    if (lastGroup.type === 'fieldValues')
      F.mergeOn(lastGroup, { groupCounts: true, skip: true })

    if (exportAllPages) _.each(_.setOn('groupCounts', true), groups)

    let cardinalityNode = await run({
      ...node,
      columns: [],
      rows: [],
      [type]: groups, // getting only groups for this type
      values: _.map(_.set('skip', true), node.values),
      expanded: {
        [type]: true,
      },
    })

    return _.get('context.results', cardinalityNode)
  }

  // Querying sequentially to reduce the load on ES
  let columnGroupingResult = await getGroupingResult('columns')
  let rowGroupingResult = await getGroupingResult('rows')

  let columnGroupingSize = getGroupingSize(
    node,
    'columns',
    columnGroupingResult,
    exportAllPages
  )
  let rowGroupingSize = getGroupingSize(node, 'rows', rowGroupingResult, exportAllPages)
  let valuesSize = _.size(node.values) || 1

  let pivotSize = columnGroupingSize * rowGroupingSize * valuesSize

  let pivot = {
    node: {
      ...node,
      context: { results: { ...columnGroupingResult, ...rowGroupingResult } },
    },
    getTotalRecords() {
      return pivotSize
    },
    async *[Symbol.asyncIterator]() {
      let search = getContextureClient(tree, {
        ...node,
        expanded: {
          columns: true,
        },
      })
      await search.refresh(['root'])

      let cNode = (pivot.node = search.getNode(['root-parent', node.key]))

      let yieldRows = async function* (drilldown) {
        let result = resultsForDrilldown(
          'rows',
          drilldown,
          _.get('context.results', cNode)
        )
        let index = 0
        let level = _.size(drilldown)

        for (let row of result.rows) {
          let path = [...drilldown, getKey(row)]
          yield {
            ...row,
            path,
            index: index++,
            level,
            recordCount: getGroupingSize(node, 'columns', row, exportAllPages) * valuesSize,
            rows: undefined, // removing children rows to avoid memory leaks
          }

          // if this row can have nested rows
          if (_.size(drilldown) < _.size(cNode.rows) - 1) {
            await cNode.expand('rows', path)

            yield* yieldRows(path)

            cNode.collapse('rows', path)
          }
        }

        // TODO pagination if exportAllPages
      }

      yield* yieldRows([])

      let totalRow = _.get('context.results', cNode)

      yield {
        ...totalRow,
        isTotalRow: true,
        path: [],
        index: 0,
        level: -1,
        recordCount: getGroupingSize(node, 'columns', totalRow, exportAllPages) * valuesSize,
        rows: undefined, // removing children rows to avoid memory leaks
      }
    },
  }
  return pivot
}
