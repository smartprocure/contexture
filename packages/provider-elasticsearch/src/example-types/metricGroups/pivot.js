let F = require('futil')
let _ = require('lodash/fp')
let { getStats } = require('./stats')
let { getField } = require('../../utils/fields')
let types = require('../../../src/example-types')
let { basicSimplifyTree } = require('../../utils/elasticDSL')

let lookupTypeProp = (def, prop, type) =>
  _.getOr(def, `${type}GroupStats.${prop}`, types)

// PivotTable -> Query:
//  rows -> columns -> values
//  (same) values for ALL group levels
// Rows/Columns are buckets, values are metrics

let aggsForValues = (node, schema) =>
  _.flow(
    // Add `pivotMetric-` to auto keys so we can skip camelCasing it in the response
    _.keyBy(
      ({ key, type, field }) =>
        key || F.compactJoin('-', ['pivotMetric', type, field])
    ),
    // This is a very interesting lint error - while key is not used in the function body, it is important.
    // Omitting key here would include it in the props spread and then pass it along as part of the body of the ES aggregation.
    // eslint-disable-next-line
    _.mapValues(({ key, type, field, ...props }) => ({
      [_.snakeCase(type)]: {
        ...props,
        ...(field && { field: getField(schema, field) }),
      },
    }))
  )(node)

let buildQuery = async (node, schema, getStats) => {
  let drilldowns = node.drilldown || []
  // Don't consider deeper levels than +1 the current drilldown
  // This allows avoiding expansion until ready
  // Opt out with falsey drilldown
  let groups = node.drilldown
    ? _.take(_.size(node.drilldown) + 1, node.groups)
    : node.groups

  let statsAggs = { aggs: aggsForValues(node.values, schema) }
  // buildGroupQuery applied to a list of groups
  let buildNestedGroupQuery = (statsAggs, groups, groupingType) =>
    _.reduce(
      async (children, group) => {
        // Subtotals calculates metrics at each group level, not needed if flattening or in chart
        // Support for per group stats could also be added here - merge on another stats agg blob to children based on group.stats/statsField or group.values
        if (node.subtotals) children = _.merge(await children, statsAggs)
        let { type } = group
        let build = lookupTypeProp(_.identity, 'buildGroupQuery', type)
        return build(group, await children, groupingType, schema, getStats)
      },
      statsAggs,
      _.reverse(groups)
    )
  if (node.columns)
    statsAggs = await buildNestedGroupQuery(statsAggs, node.columns, 'columns')
  let query = await buildNestedGroupQuery(statsAggs, groups, 'groups')

  let filters = _.compact(
    await Promise.all(
      F.mapIndexed((group, i) => {
        let filter = lookupTypeProp(_.stubFalse, 'drilldown', group.type)
        // stamp on drilldown from root if applicable
        let drilldown = drilldowns[i] || group.drilldown
        return drilldown && filter({ drilldown, ...group }, schema, getStats)
      }, groups)
    )
  )
  if (!_.isEmpty(filters))
    query = {
      aggs: {
        pivotFilter: {
          filter: { bool: { must: filters } },
          ...query,
        },
      },
    }

  return query
}

// TODO: instead of groupN, maybe look at query to say something more valuable?
//  e.g. => node.groups[n].field + ' ' + node.groups[n].type (aka 'Organization.NameState fieldValues'), maybe customized per type?
//      eg. `PO.IssuedDate month dateInterval`
// maybe client side since it's schema label driven?
let bucketToGroupN = (bucket, n) => ({
  [`group${n}`]: bucket.keyAsString || bucket.key,
})
let Tree = F.tree(_.get('groups'))
let flattenGroups = Tree.leavesBy((node, index, parents) => ({
  ...node,
  // Add groupN keys
  ..._.mergeAll(
    F.mapIndexed(
      bucketToGroupN,
      // dropping root parent (since that is just the aggs top level - would change if we add "totals" to the subtotals)
      // might also change based on what we pass in (e.g. pass in aggregations.groups?)
      _.reverse([node, ..._.dropRight(1, parents)])
    )
  ),
}))

let processResponse = (response, node = {}) => {
  let input = F.getOrReturn('pivotFilter', response.aggregations)
  // SUPER HACKY TEMPORARY METHOD
  let results = basicSimplifyTree(input)
  return { results: node.flatten ? flattenGroups(results) : results.groups }
}

let pivot = {
  aggsForValues,
  buildQuery,
  processResponse,
  validContext: node => node.groups.length && node.values.length,
  // TODO: unify this with groupStatsUtil - the general pipeline is the same conceptually
  async result(node, search, schema) {
    let query = await buildQuery(node, schema, getStats(search))
    // console.log(JSON.stringify({ query }, 0, 2))
    let response = await search(query)
    // console.log(JSON.stringify({ response }, 0, 2))
    let result = processResponse(response, node)
    // console.log(JSON.stringify({ result }, 0, 2))
    return result
  },
}
module.exports = pivot
