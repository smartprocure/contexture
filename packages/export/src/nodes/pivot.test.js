import { jest } from '@jest/globals'
import _ from 'lodash/fp.js'
import F from 'futil'
import pivot, { getKey } from './pivot.js'

let RowTree = F.tree(_.get('rows'), (key) => ({ key }))

describe('pivot', () => {
  let pivotNode = {
    key: 'report',
    type: 'pivot',
    columns: [
      {
        field: 'PO.IssuedDate',
        type: 'dateInterval',
        interval: 'year',
      },
    ],
    rows: [
      {
        field: 'Organization.State',
        type: 'fieldValues',
        filter: '',
        size: 10,
        sort: {
          field: 'count',
          direction: 'desc',
        },
        additionalFields: [],
      },
    ],
    values: [
      {
        field: 'LineItem.TotalPrice',
        type: 'sum',
      },
    ],
  }

  let defaultTree = {
    key: 'root',
    children: [
      pivotNode,
      {
        key: 'results',
        type: 'results',
      },
    ],
  }

  let pivotResult = {
    columns: [
      {
        keyAsString: '2021-01-01T00:00:00.000Z',
        key: 1609459200000,
        count: 9841186,
        'sum-LineItem.TotalPrice': 129783208652.72542,
      },
      {
        keyAsString: '2022-01-01T00:00:00.000Z',
        key: 1640995200000,
        count: 21751429,
        'sum-LineItem.TotalPrice': 215216937070.0412,
      },
      {
        keyAsString: '2023-01-01T00:00:00.000Z',
        key: 1672531200000,
        count: 7810898,
        'sum-LineItem.TotalPrice': 52196535854.69407,
      },
    ],
    rows: [
      {
        key: 'Texas',
        count: 30732145,
        metric: 314939690225.14624,
        columns: [
          {
            keyAsString: '2022-01-01T00:00:00.000Z',
            key: 1640995200000,
            count: 17255993,
            'sum-LineItem.TotalPrice': 170772861458.61346,
          },
          {
            keyAsString: '2023-01-01T00:00:00.000Z',
            key: 1672531200000,
            count: 6158859,
            'sum-LineItem.TotalPrice': 35109618846.44213,
          },
        ],
        'sum-LineItem.TotalPrice': 314939690225.14624,
      },
      {
        key: 'Florida',
        count: 8671368,
        metric: 82256991352.31442,
        columns: [
          {
            keyAsString: '2021-01-01T00:00:00.000Z',
            key: 1609459200000,
            count: 2523893,
            'sum-LineItem.TotalPrice': 20725998732.634754,
          },
          {
            keyAsString: '2022-01-01T00:00:00.000Z',
            key: 1640995200000,
            count: 4495436,
            'sum-LineItem.TotalPrice': 44444075611.42773,
          },
          {
            keyAsString: '2023-01-01T00:00:00.000Z',
            key: 1672531200000,
            count: 1652039,
            'sum-LineItem.TotalPrice': 17086917008.251938,
          },
        ],
        'sum-LineItem.TotalPrice': 82256991352.31442,
      },
    ],
    'sum-LineItem.TotalPrice': 397196681577.4607,
    count: 39403513,
  }

  let getService = () =>
    jest.fn((tree) => {
      _.last(tree.children).context = { results: pivotResult }
      return tree
    })

  let prepareSimpleStrategy = async () =>
    pivot({
      service: getService(),
      tree: _.cloneDeep(defaultTree),
      ...pivotNode,
    })

  it('should retrieve estimated export credits', async () => {
    let strategy = await prepareSimpleStrategy()
    expect(await strategy.getTotalRecords()).toBe(12)
  })
  it('retrieves records', async () => {
    let service = getService()
    let strategy = await prepareSimpleStrategy()

    let expectedRecords = RowTree.toArrayBy(
      (record, index = 0, parents) => ({
        ...record,
        index,
        level: parents.length - 1,
        path: _.compact([getKey(record)]),
        recordCount: _.size(record.columns) + 1,
      }),
      pivotResult
    )
    expectedRecords.push({ ...expectedRecords.shift(), isTotalRow: true })

    let records = []
    for await (const r of strategy) records.push(r)

    let exportedRecordCount = _.reduce(
      (acc, record) => acc + record.recordCount,
      0,
      records
    )

    expect(exportedRecordCount).toEqual(11)
    expect(records).toEqual(expectedRecords)
    expect(service).toMatchSnapshot()
  })
})
