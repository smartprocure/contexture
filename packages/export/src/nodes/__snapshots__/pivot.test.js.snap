// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pivot retrieves records 1`] = `[MockFunction]`;
